### PR TITLE
Add better crash dialog

### PR DESCRIPTION
**Pull request by NathanW2**
## Description

Adds a nicer crash dialog for better UX. Crashing sucks so best to handle it well if we can. Showing a crash dumped dialog before was a bit raw so this is attempt at doing something a bit nicer.

@nirvn can I have a cool "sad" QGIS logo :) 

Things to do:

- [ ] Funny error message
- [ ] Cool crash image
- [x] Wire up buttons
- [x] Write dev guide (maybe we should just have this link to online someplace)

image

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -4724,6 +4725,8 @@ void QgisApp::fileExit()
 
 void QgisApp::fileNew()
 {
+    QgsRuntimeProfiler* profile;
+    profile->clear();
```
Comment (NathanW2) [inline src/app/qgisapp.cpp]: Ignore this. Only here for a quick way to make it crash.

**Comment by nyalldawson** [comment]
Can we put the mantra in here? I doubt there's any spam risk...

**Comment by NathanW2** [comment]
@rduivenvoorde thoughts on just routing "Keen to help" link to the docs site?  Might be better then just trying to shove it into the UI.

**Comment by m-kuhn** [comment]
Oh, nifty.

Linking the docs site is the proper approach - or an integrated error-reporter that will upload the stack trace and has a text field where the user can enter a few sentences what he has been doing.
I'd also be in favor of showing the mantra here.

**Comment by rduivenvoorde** [comment]
@NathanW2 yep, I would link to one of the 'getinvolved' pages, like:

http://qgis.org/en/site/getinvolved/development/bugreporting.html#bugs-features-and-issues

or write a FAQ item (linking to that page): http://qgis.org/en/site/getinvolved/faq/index.html

@m-kuhn I would not been to sure about the mantra, it would make it harder to change it and with currenlty >3000 open issues, at about 5-10 coming in every day, I'm not sure if I would be to happy with even more 'fire and run'-issuers...

**Comment by m-kuhn** [comment]
@rduivenvoorde the nice thing about this one would be that they include a trace - which makes them a) very high quality and b) easy to aggregate duplicates. So much better than any other report.

**Comment by rduivenvoorde** [comment]
@m-kuhn Ok, but only nightlies and weeklies have debuginfo isn't it? Or is there also a way to extract such traces from windows Release builds? I'm off course ok with adding the stacktrace to it anyway (if possible).
It is just that I'm not to confident that non techies are determined enough to follow up (and I'm not in favour of releasing the mantra other then via a trust relationship).

**Comment by NathanW2** [comment]
Let's deal with the whole getting a useful stack trace in a different PR. I
suspect it's going to take more work.  I will just keep this one as a
simple nicer message for now and we can add more options later.

**Comment by m-kuhn** [comment]
@NathanW2 totally agree, that's not a requirement for this PR.

**Comment by NathanW2** [comment]
Never was ;)  But is still on my todo if we can work out how.

**Comment by tomchadwin** [comment]
Re: Funny error message

Including the word "sorry" in the dialog might also soften the impact of a crash. Having Chrome say "Aw, snap!" to the audience at the end of my FOSS4GUK qgis2web talk got a laugh.

**Comment by NathanW2** [comment]
@tomchadwin yep open to rewording that error message. Post any ideas that you have.

**Comment by dakcarto** [comment]
"Oh, snapping tolerance!" (just kind of kidding)

**Comment by nyalldawson** [comment]
My vote is for "error 999999(+1)" ;)

**Comment by NathanW2** [comment]
@dakcarto that is very tempting. Very tempting indeed.

File: src/app/qgisapp.cpp [hunk 2]
```diff
@@ -12491,42 +12492,53 @@ void QgisApp::transactionGroupCommitError( const QString &error )
...
-    }
-    else
-    {
-      msg = QObject::tr( "writing of minidump to %1 failed (%2)" ).arg( dumpName ).arg( GetLastError(), 0, 16 );
-    }
-
-    CloseHandle( hDumpFile );
-  }
-  else
+  // Crash dump creation will be move to a new class in the near future.
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: Can you #ifdef this out instead?

**Review by nyalldawson — APPROVED** [review]

File: src/ui/qgscrashdialog.ui [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+       </property>
+       <property name="text">
+        <string>Ouch!</string>
+       </property>
+      </widget>
+     </item>
+     <item>
+      <widget class="QLabel" name="label_3">
+       <property name="text">
+        <string>Looks lke QGIS has crashed. Something unexpected happened that we didn't handle correctly</string>
```
Comment (nyalldawson) [inline src/ui/qgscrashdialog.ui]: lke -> like

Also needs a trailing "."

**Comment by nyalldawson** [comment]
Any way we could move the buttons to the bottom of the dialog instead of near the middle?

**Comment by m-kuhn** [comment]
Looking forward to seeing this translated literally :P 

Where has the "help us fix this" link gone? That's also useful without any stacktrace etc. attached

**Comment by NathanW2** [comment]
It's still there just in code now.  Links out to the website about bug
reports.

**Comment by rduivenvoorde** [comment]
Hi Gents, about the catchy message ++1 nicely found, but don't want to play Grumpy here, but during a course I had a lot of processing algorithm crashes, and during that stress the 'funny message' from there was not funny at all anymore after 2 times :-(
And the combi of crashes + funny 'oopsy...'s did not make a very good impression at that moment.

So it is too much to maybe do the funny message twice (during a week), and then move on to a more humble one "sorry guys.... bla bla'?

Should the second sentence "Something..." not be on a new line or paragraph? (/me not so into this...)

**Comment by NathanW2** [comment]
@rduivenvoorde if you are getting a crash it's never funny.  Main thing is you shouldn't be seeing this message in the first place so no matter what we put here will make people annoyed.   We can be funny but serious at the same time.

**Comment by NathanW2** [comment]
Merged in 7694dd2d8

**Comment by NathanW2** [comment]
We can update the image and text later but at least the dialog is nicer now.

**Pull request closed without merging**